fix: allow `Require Linear History` when selecting merge method

tpolekhin opened the pull request:
## what
Account for branch protection rule "Require Linear History" when choosing merge method
<!--
- Describe high-level what changed as a result of these commits (i.e. in plain-english, what do these changes mean?)
- Use bullet points to be concise and to the point.
-->


## why
When repo config allows merge commits:
<img width="797" alt="Screenshot 2023-03-18 at 13 40 13" src="https://user-images.githubusercontent.com/10550832/226106700-06a35e14-497c-482f-94c4-3fe13bca3f8d.png">

But branch protection rules requires linear history (aka banning merge commits):
<img width="506" alt="Screenshot 2023-03-18 at 13 40 36" src="https://user-images.githubusercontent.com/10550832/226106733-0b6a99b0-e3de-404e-9389-ffd86fdaaabe.png">

Current logic will always pick `merge` method by default. This PR fixes that.

<!--
- Provide the justifications for the changes (e.g. business case). 
- Describe why these changes were made (e.g. why do these commits fix the problem?)
- Use bullet points to be concise and to the point.
-->

## tests
- [x] I have updated [TestGithubClient_MergePullCorrectMethod](https://github.com/runatlantis/atlantis/blob/main/server/events/vcs/github_client_test.go#L832) test
- [x] I have tested my changes by running tests from VSCode
- [x] I have tested my changes by running `make test`
- [x] I have tested my changes on test environment

<!--
- [ ] I have tested my changes by ...
-->

## references

<!--
- Link to any supporting github issues or helpful documentation to add some context (e.g. stackoverflow). 
- Use `closes #123`, if this PR closes a GitHub issue `#123`
-->
- Closes #1176

nitrocode commented on server/events/vcs/github_client.go:
```diff
@@ -579,13 +579,21 @@ func (g *GithubClient) MergePull(pull models.PullRequest, pullOptions models.Pul
 	if err != nil {
 		return errors.Wrap(err, "fetching repo info")
 	}
+	protection, resp, err := g.client.Repositories.GetBranchProtection(context.Background(), *repo.Owner.Name, *repo.Name, pull.BaseBranch)
```
If git branch protections aren't set, will this pass? Do we need to check for branch protections first before callign this function?

tpolekhin replied:
According to the GitHub api docs it will return 404 if branch protection is not set for the requested branch

nitrocode commented on server/events/vcs/github_client.go:
```diff
@@ -579,13 +579,21 @@ func (g *GithubClient) MergePull(pull models.PullRequest, pullOptions models.Pul
+	if err != nil {
+		return errors.Wrap(err, "getting branch protection rules")
+	}
+	requireLinearHistory := false
```
Please add a test for these changes

tpolekhin commented:
@nitrocode I have fixed the logic a little bit and also added test. `deploy/netlify` fails because of some frontend issue I think after the recent renovate deps update - I didn't touch the frontend part in my PR.
Could you please approve test workflows here, or you want to wait until the frontend issue is fixed and then rebase this PR?

nitrocode commented:
Please ignore the netlify error and please fix the tester error.

tpolekhin commented:
> Please ignore the netlify error and please fix the tester error.

@nitrocode do you mean `tester / website_link_check`? I think this one is also affected by the frontend issue

nitrocode commented:
Yes please ignore the netlify error. Seems to have come from this pr https://github.com/runatlantis/atlantis/pull/3225

nitrocode commented on server/events/vcs/testdata/github-branch-protection-require-linear-history.json:
```diff
@@ -0,0 +1,6 @@
+    "required_linear_history": {
+        "enabled": true
+    }
+}
```
Mind adding a new line here? We probably need a gitattributes in this repo. If you can add one too for .go and .json that would be nice too

tpolekhin replied:
IDK why but VSCode automatically removes a newline on save for the json file
so I guess this is a "accepted pattern"?

nitrocode replied:
No worries. We can automate this away using https://github.com/runatlantis/atlantis/pull/3244

nitrocode commented on server/events/vcs/github_client.go:
```diff
@@ -579,13 +579,23 @@ func (g *GithubClient) MergePull(pull models.PullRequest, pullOptions models.Pul
+	protection, _, err := g.client.Repositories.GetBranchProtection(context.Background(), repo.Owner.GetLogin(), *repo.Name, pull.BaseBranch)
+	if err != nil {
+		if !errors.Is(err, github.ErrBranchNotProtected) {
+			return errors.Wrap(err, "getting branch protection rules")
```
How come we want to do an early return here?

tpolekhin replied:
We check if the branch is protected by calling the API endpoint.
If branch is protected we get back the response and no error.
If branch is not protected we get back nil response and specific error message.
All other cases we treat as standard communication errors 401, 504, 500 etc.

nitrocode commented on server/events/vcs/github_client.go:
```diff
@@ -579,13 +579,23 @@ func (g *GithubClient) MergePull(pull models.PullRequest, pullOptions models.Pul
 	if err != nil {
 		return errors.Wrap(err, "fetching repo info")
 	}
+	protection, _, err := g.client.Repositories.GetBranchProtection(context.Background(), repo.Owner.GetLogin(), *repo.Name, pull.BaseBranch)
```
Could you test this pr in your own setup to confirm linear history enabled and linear history disabled work as expected?

nitrocode commented:
Please fill out the pr template

tpolekhin commented:
@nitrocode I have tested this change on the actual GitHub repository
https://github.com/tpolekhin/atlantis-test-terraform

- `main` branch - unprotected - automerge successful after `atlantis apply`
- `protected` branch - protected without linear history - successful
- `protected-linear` branch - protected WITH linear history - successful

Atlantis logs: https://gist.github.com/tpolekhin/eb63628798b82e7584d3f2eac14dcd16

nitrocode commented:
@tpolekhin great work and testing! Please feel free to propose other PRs if you have the time. And please join the #contributors channel in the atlantis slack community. Thank you!

tpolekhin commented:
@nitrocode thank you for the great feedback and for the merge!

nitrocode commented:
@tpolekhin this was reverted due to a breaking change. See the related pr https://github.com/runatlantis/atlantis/pull/3321
